data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/router/DataFlowComponentRouter.java

```diff
@@ -5,81 +5,41 @@
 import java.util.function.BiConsumer;
 
 /**
  * Package-protected utility to route for a single {@link DataFlowComponent}. This is
  * intended to help break apart {@link Router} for better testing.
  */
 class DataFlowComponentRouter {
-    Record getRecordForComponent(Set<Record> usedRecords, Record record) {
-        if (usedRecords == null) {
-            return record;
-        }
-        if (!usedRecords.contains(record)) {
-            usedRecords.add(record);
-            return record;
-        }
-        if (record.getData() instanceof JacksonSpan) {
-            try {
-                final Span spanEvent = (Span) record.getData();
-                Span newSpanEvent = JacksonSpan.fromSpan(spanEvent);
-                return new Record<>(newSpanEvent);
-            } catch (Exception ex) {
-            }
-        } else if (record.getData() instanceof Event) {
-            try {
-                final Event recordEvent = (Event) record.getData();
-                Event newRecordEvent = JacksonEvent.fromEvent(recordEvent);
-                return new Record<>(newRecordEvent);
-            } catch (Exception ex) {
-            }
-        }
-        return record;
-    }
-
     <C> void route(final Collection<Record> allRecords,
                    final DataFlowComponent<C> dataFlowComponent,
                    final Map<Record, Set<String>> recordsToRoutes,
-                   Set<Record> usedRecords,
+                   final RouterGetRecordStrategy getRecordStrategy,
                    final BiConsumer<C, Collection<Record>> componentRecordsConsumer) {
 
         final Collection<Record> recordsForComponent;
         final Set<String> dataFlowComponentRoutes =  dataFlowComponent.getRoutes();
 
         if (dataFlowComponentRoutes.isEmpty()) {
-            if (usedRecords == null || usedRecords.isEmpty()) {
-                recordsForComponent = allRecords;
-                if (usedRecords != null) {
-                    usedRecords.addAll(allRecords);
-                }
-            } else {
-                recordsForComponent = new ArrayList<>();
-                for (Record record : allRecords) {
-                    recordsForComponent.add(getRecordForComponent(usedRecords, record));
-                }
-            }
+            recordsForComponent = Objects.isNull(getRecordStrategy) ? allRecords : getRecordStrategy.getAllRecords(allRecords);
         } else {
             recordsForComponent = new ArrayList<>();
             for (Record record : allRecords) {
                 final Set<String> routesForEvent = recordsToRoutes
                         .getOrDefault(record, Collections.emptySet());
 
                 if (routesForEvent.stream().anyMatch(dataFlowComponentRoutes::contains)) {
-                    recordsForComponent.add(getRecordForComponent(usedRecords, record));
+                    recordsForComponent.add(Objects.isNull(getRecordStrategy) ? record : getRecordStrategy.getRecord(record));
```

Comment:
I think we should require the strategy. It will simplify the code and allow you to have a consistent set of tests.